relayer/src/chain/cosmos.rs

```diff
@@ -373,13 +379,65 @@ impl CosmosSdkChain {
         let mut tx_bytes = Vec::new();
         prost::Message::encode(&tx_raw, &mut tx_bytes).unwrap();
 
-        let response = self.block_on(broadcast_tx_sync(self, tx_bytes))?;
+        self.block_on(broadcast_tx_sync(self, tx_bytes))
+    }
+
+    // Try to send_tx with retry on account sequence error.
+    // An account sequence error can occur if the cached account sequence of
+    // the relayer somehow get outdated, or when the relayer's wallet is used
+    // concurrently elsewhere.
+    // When there is an account sequence mismatch, we refetch the account sequence
+    // from the chain and retry sending the transaction again with the new sequence.
+    fn send_tx_with_account_sequence_retry(
+        &mut self,
+        proto_msgs: Vec<Any>,
+        retries: u32,
+    ) -> Result<Response, Error> {
+        let account_sequence = self.account_sequence()?;
+
+        let response = self.send_tx_with_account_sequence(proto_msgs.clone(), account_sequence)?;
 
         debug!("[{}] send_tx: broadcast_tx_sync: {:?}", self.id(), response);
 
-        self.incr_account_sequence()?;
+        match response.code {
+            Code::Ok => {
+                self.incr_account_sequence();
+                Ok(response)
+            }
+            // Cosmos SDK defines an account sequence mismatch with the error code 32
+            Code::Err(32) => {
+                if retries < MAX_ACCOUNT_SEQUENCE_RETRY {
+                    warn!("send_tx failed with incorrect account sequence. retrying with account sequence refetched from the chain.");
+
+                    // Refetch account information from the chain.
```

Comment:
I appreciate the thorough comments! Also, the solution of refreshing only on error is simpler than what I had in mind (parsing error message) and probably more future-proof. Cool work Soares!

I have two concerns:

- were you able to reproduce the error, and could you describe the scenario? Is it as simple as using two instances of hermes, a `start` and an `ft-transfer` concurrently, or did you stumble upon more advanced scenarios that we should be aware while testing?
- seems like a tail-recursive method call, but the max depth is `5` so not too big. My shallow understanding of this is that it shouldn't pose problems from a memory-use perspective, so I just wanted to run it by you.

Reply:
> seems like a tail-recursive method call, but the max depth is 5 so not too big.

It shouldn't be an issue, and usually the retry should succeed after one try anyway. The only time when the max retry would exceed is if the relayer is configured with a shared wallet that is being used concurrently all the time, which in that case is considered a misconfiguration.